geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+    val INT_AS_LONG = bind[Int](_.writeLong(_), _.readLong().toInt,  Schema.create(Schema.Type.LONG))
+    implicit val LONG = bind[Long](_.writeLong(_), _.readLong(),  Schema.create(Schema.Type.LONG))
+    val FLOAT_AS_LONG = bind[Float]((e,l)=>e.writeLong(l.toLong), _.readLong(),  Schema.create(Schema.Type.LONG))
+    val DOUBLE_AS_LONG = bind[Double]((e,l)=>e.writeLong(l.toLong), _.readLong(),  Schema.create(Schema.Type.LONG))
+
+    val BYTE_AS_FLOAT = bind[Byte](_.writeFloat(_), _.readFloat().toByte,  Schema.create(Schema.Type.FLOAT))
+    val SHORT_AS_FLOAT = bind[Short](_.writeFloat(_), _.readFloat().toShort,  Schema.create(Schema.Type.FLOAT))
+    val INT_AS_FLOAT = bind[Int](_.writeFloat(_), _.readFloat().toInt,  Schema.create(Schema.Type.FLOAT))
+    val LONG_AS_FLOAT = bind[Long](_.writeFloat(_), _.readFloat().toLong,  Schema.create(Schema.Type.FLOAT))
+    implicit val FLOAT = bind[Float](_.writeFloat(_), _.readFloat(),  Schema.create(Schema.Type.FLOAT))
+    val DOUBLE_AS_FLOAT = bind[Double]((e,f)=>e.writeFloat(f.toFloat), _.readFloat(),  Schema.create(Schema.Type.FLOAT))
+
+    val BYTE_AS_DOUBLE = bind[Byte](_.writeDouble(_), _.readDouble().toByte,  Schema.create(Schema.Type.DOUBLE))
+    val SHORT_AS_DOUBLE = bind[Short](_.writeDouble(_), _.readDouble().toShort,  Schema.create(Schema.Type.DOUBLE))
+    val INT_AS_DOUBLE = bind[Int](_.writeDouble(_), _.readDouble().toInt,  Schema.create(Schema.Type.DOUBLE))
+    val LONG_AS_DOUBLE = bind[Long](_.writeDouble(_), _.readDouble().toLong,  Schema.create(Schema.Type.DOUBLE))
+    val FLOAT_AS_DOUBLE = bind[Float](_.writeDouble(_), _.readDouble().toFloat,  Schema.create(Schema.Type.DOUBLE))
+    implicit val DOUBLE = bind[Double](_.writeDouble(_), _.readDouble(),  Schema.create(Schema.Type.DOUBLE))
+  }
+
+  trait DerivedCodecs extends PrimitiveCodecs {
+    def derive[T,U](encMethod: T=>U, decMethod: U=>T)(implicit primitiveCodec: AvroCodec[U], tag: ClassTag[T]) = {
+      val codec = new AvroCodec[T](primitiveCodec.schema) {
+        override val clazz = tag.runtimeClass.asInstanceOf[Class[T]]
+        override def enc(t: T, out: Encoder) = primitiveCodec.enc(encMethod(t), out)
+        override def dec(in: Decoder) = decMethod(primitiveCodec.dec(in))
+      }
+      registry.getOrElseUpdate(tag.runtimeClass, mutable.LinkedHashMap.empty)(primitiveCodec.schema.getType)=codec
+      codec
+    }
+
+    implicit val DATE = derive[Date,Long](_.getTime, new Date(_))
+    implicit val CALENDAR = derive[Calendar,Long](_.getTimeInMillis, d=>{val gc=new GregorianCalendar(); gc.setTimeInMillis(d); gc})
+    implicit val SQL_DATE = derive[sql.Date,Long](_.getTime, new sql.Date(_))
+    implicit val SQL_TIMESTAMP = derive[Timestamp,Long](_.getTime, new Timestamp(_))
+    implicit val _UUID = derive(encodeUUID, decodeUUID)
+    implicit val URI = derive[net.URI,String](_.toString, net.URI.create)
+  }
+
+  class StatefulCodecs(wkbWriter: WKBWriter, wkbReader: WKBReader) extends DerivedCodecs {

```

Comment:
Other places in the GeoMesa codebase, we use object pools or ThreadLocals to prevent issues with these classes not being thread-safe. 

If we try to address concurrency, that might be useful.